### ISSUE TITLE
python 3 support

### ISSUE THREAD

**Issue by BurntSushi** [BurntSushi/nflgame#78]
I think it's about time we have an issue for this. At some point, we need to press forward. I have a small hiatus coming up that I could use to push us to Python 3.

The good news is that all dependencies (pytz, beautifulsoup4, httplib2 and psycopg2 (for nfldb)) are supported on Python 3. So the only task is to make the actual nflgame module Python 3 compatible.

There are two ways forward. The first is to migrate to Python 3 and leave Python 2 in the wind. The second is to simultaneously support Python 2 and 3 by writing a subset of each that is compatible with both versions.

Writing Python 2/3 compatible code is difficult. I've done it with [pdoc](https://github.com/BurntSushi/pdoc) and it is not fun. Also, I don't find the prospect of debugging user problems with Python 2/3 code very enticing.

Writing code that is only compatible with Python 3 is much simpler and easier to maintain. However, it makes the library unusable to people who cannot run Python 3. It will also force everyone to upgrade their personal scripts that use nflgame to Python 3. That's a major bummer.

**Comment by digitalmonarch** [comment]
Wanted to throw my hat in the ring too. I'm in favor of moving to Python 3. I'm new to python but not new to software development, and I would like to start by building some stuff that uses nflgame. I'm hesitant to learn on 2.7 given its age and some of the quirks that I've read about. It's hard to get motivated to learn on something that's deprecated :)

**Comment by BurntSushi** [comment]
@digitalmonarch Python 2.7 isn't deprecated. It's being supported until at least 2020, I think.

FWIW, most of the world is still using 2.7. The split is unfortunate, but it's not nearly as bad as you might think. The differences between 2.7 and 3.x are pretty small, but they are subtle enough to be a major pain. Otherwise, your day-to-day programming life isn't going to be impacted too much whether you're on 2 or 3.

**Comment by Croxy** [comment]
I'm curious, couldn't you fork your code and have two separate platforms? You could slowly start moving support over to Python 3... I don't know, I just found this repo today and am looking forward to playing with it!

**Comment by BurntSushi** [comment]
Maintaining two separate code bases is absolutely insane.

**Comment by Croxy** [comment]
Haha, fair enough. I was thinking you could EOL Python 2 support.

**Comment by BurntSushi** [comment]
Ah i see. Yes, that is more reasonable.

**Comment by sayeghr** [comment]
Perhaps it would be best if we were to start writing unit tests for nflgame until we get a large % of lines covered by our test code. Once we do that, we can start running the tests in Python 2 and in Python 3 and fixing the code base to be compatible with both versions.

**Comment by BurntSushi** [comment]
@sayeghr Yes, that would be nice. One might start with porting these tests from nfldb: https://github.com/BurntSushi/nfldb/blob/master/tests/test_query.py

**Comment by mlenzen** [comment]
FYI, I've ported nflgame to Python3: https://github.com/mlenzen/nflgame

It runs without error in 2.7.9 and 3.4.3 but I haven't compared outputs (yet) or written tests.

I also switched to requests instead of httplib2 because I was familiar with it and it handles content decoding and I didn't want to think about it.

**Comment by jsheedy** [comment]
I only started in with nflgame today, but wanted to give my vote for Python3. I have an existing tool which is dependent on the new async in Python3 that I'm hoping to try integrating nflgame into. I agree with the comments that maintaining support for 2 and 3 might be untenable, and a clean break in a major release might be an OK solution. I haven't tried the fork from @mlenzen yet, but I'll give it a shot. Thanks for all the great work!

**Comment by BurntSushi** [comment]
This really isn't a matter of voting. It's a matter of developing a test suite and being confident that the port works. Thank you @mlenzen for getting it started!

**Comment by MarkRConway** [comment]
You can go to Python 3 now. Make a break for it and force others to leave 2.X in their wake. Thanks for the awesome package!

**Comment by Irish1986** [comment]
As a newcomer to coding and python in general, what are the realistic option if I desire to use Py3with NFLGame ?

Should I clone the project and run something like 2to3 or six on my own branch ?  How can I make the work I do beneficial for the greater good ?  I don't really want to get into Py2 now and there are so many other cool Py3 package I wish to use.

**Comment by mlenzen** [comment]
@Irish1986 I have a working python3 fork at https://github.com/mlenzen/nflgame

**Comment by derek-adair** [comment]
Whats he status on this? This is a must have for me.

**Comment by BurntSushi** [comment]
Nothing to report. I'm unlikely to work on this.

**Comment by derek-adair** [comment]
I've figured out a way around it for now... But it would simplify things for me in the long run.

What would it take to get  https://github.com/mlenzen/nflgame merged in?
